consul-lambda-registrator/proxy/proxy.go

```diff
@@ -0,0 +1,112 @@
+package proxy
+
+import (
+	"errors"
+	"fmt"
+	"sync"
+	"sync/atomic"
+)
+
+// Server implements a proxy server that manages TCP listeners for a configurable set of upstreams.
+type Server struct {
+	cfgs      []*Config
+	listeners []*Listener
+
+	// waitChan is closed once the server is up and running. It can be used by
+	// callers to wait until the server is initialized and ready to handle connections.
+	waitChan chan struct{}
+
+	stopFlag int32
+	stopChan chan struct{}
+}
+
+// New returns a new, unstarted proxy server from the given proxy configurations.
+// The proxy can be started by calling Serve.
+func New(cfgs ...*Config) *Server {
+	return &Server{
+		waitChan: make(chan struct{}),
+		stopChan: make(chan struct{}),
+		cfgs:     cfgs,
+	}
+}
+
+// Serve starts the proxy and initializes all of the configured listeners.
+// It blocks until Close is called or an error occurs.
+func (s *Server) Serve() error {
+	if atomic.LoadInt32(&s.stopFlag) != 0 {
+		return errors.New("serve called on a closed server")
+	}
+
+	errChan := make(chan error)
+	lwg := &sync.WaitGroup{}
+
+	s.listeners = make([]*Listener, 0, len(s.cfgs))
+	for _, lc := range s.cfgs {
+		l := NewListener(lc)
+		s.listeners = append(s.listeners, l)
+
+		// Start the listener. If Serve returns an error it is written to the errChan and handled below.
+		go func(l *Listener) {
+			if err := l.Serve(); err != nil {
+				errChan <- fmt.Errorf("failed to serve listener: %w", err)
+				return
+			}
+		}(l)
+
+		// Add a wait for this listener to start.
+		lwg.Add(1)
+		go func(wg *sync.WaitGroup, l *Listener) {
+			defer wg.Done()
+			l.Wait()
+		}(lwg, l)
+
+		// Watch for errors on this listener.
+		go func(l *Listener) {
+			for le := range l.Errors() {
+				errChan <- le
+			}
+		}(l)
```

Comment:
Yes, I originally had the listener sending to the proxy's errChan, but this setup imposes a usage pattern on Listener.Serve in that the caller must provide an errChan whether or not they are interested in checking for errors from the listener.

I'm looking over this again now and I remember why I did it this way. The `sendError` func is to capture errors from `Listener.handleConn`, not `Listener.Serve`. The distinction is that if `Listener.Serve` returns an error, this is fatal to the proxy because it means that there is no listener for the particular upstream. However, errors returned from `handleConn` should not be fatal because they represent an error for a single connection stream (e.g. an HTTP request). I don't think we want to kill the proxy on a connection error because that would mean that retries wouldn't work. Maybe the upstream is not reachable for an instant, but then comes back.

Reply:
That makes sense.

I may be misreading / misunderstanding, but currently it looks like all Listener errors _are_ funneled back into this proxy's `errChan`? Basically, this bit on L64: https://github.com/hashicorp/terraform-aws-consul-lambda-registrator/blob/711394225446ecca320325d989ef504ccdb97a3d/consul-lambda-registrator/proxy/proxy.go#L63-L68

Does that mean that Proxy will stop when there is any error from an individual connection?

Reply:
Yes you're absolutely correct. I just pushed a commit that discriminates between the two cases. It stops on errors from `Serve` and logs errors from `handleConn`.